nemo/collections/tts/data/vocoder_dataset.py

```diff
@@ -175,28 +212,218 @@ def __getitem__(self, index):
+        shard_strategy: str = "scatter",
+        global_rank: int = 0,
+        world_size: int = 2,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if len(kwargs) > 0:
+            logging.warning(
+                f"Arguments {kwargs.keys()} does not support for TarredVocoderDataset, they will be ignored."
+            )
+
+        self.sample_rate = sample_rate
+        self.n_samples = n_samples
+
+        if trunc_duration:
+            self.trunc_samples = int(trunc_duration * self.sample_rate)
+        else:
+            self.trunc_samples = None
 
-        for example in batch:
-            dataset_name_list.append(example["dataset_name"])
-            audio_filepath_list.append(example["audio_filepath"])
-            audio_list.append(example["audio"])
-            audio_len_list.append(example["audio_len"])
+        if feature_processors:
+            logging.info(f"Found feature processors {feature_processors.keys()}")
+            self.feature_processors = list(feature_processors.values())
+        else:
+            self.feature_processors = []
 
-        batch_audio_len = torch.IntTensor(audio_len_list)
-        audio_max_len = int(batch_audio_len.max().item())
+        self.data_samples = []
+        self.audio_tar_filepaths = []
+        for dataset_name, dataset_info in dataset_meta.items():
+            audio_tar_filepaths = dataset_info.audio_tar_filepaths
+            self.audio_tar_filepaths += [audio_tar_filepaths]
+            dataset = DatasetMeta(**dataset_info)
+            samples, _ = preprocess_manifest(
+                dataset_name=dataset_name, dataset=dataset, min_duration=min_duration, max_duration=max_duration,
+            )
+            self.data_samples += samples
 
-        batch_audio = stack_tensors(audio_list, max_lens=[audio_max_len])
+        self.file_id_to_sample_map = {}
+        for sample in self.data_samples:
+            file_id = os.path.splitext(os.path.basename(sample.manifest_entry["audio_filepath"]))[0]
+            if file_id not in self.file_id_to_sample_map:
+                self.file_id_to_sample_map[file_id] = sample
+            else:
+                raise ValueError(
+                    f"Duplicate file_id {file_id} found in manifest {sample.manifest_entry['audio_filepath']}"
+                )
+
+        logging.info(f"world size: {world_size}")
+        audio_tar_filepaths = expand_sharded_filepaths(
+            sharded_filepaths=audio_tar_filepaths,
+            global_rank=global_rank,
+            world_size=world_size,
+            shard_strategy=shard_strategy,
+        )
 
-        batch_dict = {
-            "dataset_names": dataset_name_list,
-            "audio_filepaths": audio_filepath_list,
-            "audio": batch_audio,
-            "audio_lens": batch_audio_len,
+        self._dataset = wd.WebDataset(audio_tar_filepaths, nodesplitter=None)
+
+        if shuffle_n > 0:
+            self._dataset = self._dataset.shuffle(shuffle_n)
```

Comment:
Should we do something like` dataset.shuffle(size=shuffle_n, initial=shuffle_n)`? The initial buffer size of 100 is really small, which I saw slow convergence a lot on my small test datasets (but might be irrelevant on very large datasets).

Reply:
If the set is small then yes number of samples per worker would become small and `shuffle` might not have an effect and wouln;t be a problem for larger sets as number of samples per worker would be large. As per [this](https://github.com/webdataset/webdataset/issues/71) discussion, it looks like if we want to get a true shuffle of over all sets, its better to unbatch() on loader and shuffle with a larger number.